### ISSUE TITLE
Adding resources into output of bokeh.embed.components

### ISSUE THREAD

**Issue by chdoig** [bokeh/bokeh#2149]
I would like the output of `bokeh.embed.components` (or maybe, have an alternative way) to get **all** the html tags needed to visualize a Bokeh plot. 

Right now, _components assumes that BokehJS has already been loaded, for instance either inline in the document text, or from CDN_, but those resources can actually be passed to `components(plot_object, resources)`.

This makes the user have to worry about adding the same resource information in two different places, into components and into the actual template where the output of the components will be passed. Having the user pass the resources in two different places can be seen redundant and maybe be prone to errors, e.g., the user might pass a different resources into components than what is loaded in the html file where those divs and scripts are passed.

These thoughts have come after trying to reply to @mrocklin question in SO:
http://stackoverflow.com/questions/29462117/supplemental-javascript-for-bokeh-plot
_What is all of the other stuff that I need in order to embed this in a web page? I prefer to link out to externally hosted javascript rather than inlining everything._

I recommended him to use `bokeh.embed.components` with CDN, e.g.:

http://bokeh.pydata.org/en/latest/docs/user_guide/embedding.html#components

```
from bokeh.plotting import figure
from bokeh.resources import CDN
from bokeh.embed import components

plot = figure()
plot.circle([1,2], [3,4])

script, div = components(plot, CDN)
```

but then I realized that he also needed to add the CDN script tags to load Bokeh and that information was not clear or easy to find in the docs (e.g. link to cdn and what scripts actually need to be loaded: css + js or just js...), but the information is actually already available in the `CDN.js_files` and `CDN.css_files`, and that a template already exists to add them `resources.html`

**Comment by bryevdv** [comment]
Hi @chdoig I agree that the duplication is not good. The `resources` arg is only there at all for a tedious technical reason. My first preference would be to actually remove the need to supply any resources at all to `components`. What do you think about this proposal:
* remove `resources` arg from `components` 
* have a simple way to get actual `<script>` etc tags from a `Resources` object
* flesh out a full example showing these put together in the `components` docs

My preference here is to preserve a loose coupling between things when possible. 

If it is not possible to remove the `resources` arg, then I agree that `components` should be modified to return templatable resource tags as well. 

@mattpap what do you think about removing all the `-dev` and `js_wrapper` bits? It is complicating the user interface in an unnecessary way and ultimately we have to design interface for users, not for ourselves. Given that the browserify refactor broke splitjs operation (unless I am mistaken?) I think this is a good opportunity to streamline and simplify some things.

**Comment by bryevdv** [comment]
@birdsarah I believe recent work on `components` addresses all of this, do you agree?

**Comment by birdsarah** [comment]
@bryevdv, I don't think so. I'm a little confused by this issue:

It says:

```python
script, div = components(plot, CDN)
```
But if you put this in you would actually get a [warning](https://github.com/bokeh/bokeh/blob/master/bokeh/embed.py#L144) because resources argument isn't used and should be `None`.

The correct way to call components is 

```python
script, div = components(plot)
```

So from the OP:

> This makes the user have to worry about adding the same resource information in two different places, 

is not correct.

----

However, I do think there is an interesting feature request it's basically somewhere between `components` and `file_html`

`components` just gives you the script to make the plot. It doesn't give you bokehjs or the css.

`file_html` internally renders the resources (the new version I'm working on) into template variables `{{ bokeh_js }}` `{{ bokeh_css }}` and renders the template. Since the introduction of custom variables and template and the new resources stuff, it's becoming a pretty useful method for more than just very basic rendering.

What @chdoig is asking for, I think, is the following:
````python
bokeh_js, bokeh_css, script, div = all_components(plot, Resources())
````

I can see why this would have been useful in April, but it seems less useful given the current power of `file_html` and what you'd likely do with the output of `all_components`.

Having said that, this makes me inclined to open a new issue for something similar to @bryevdv's suggestion on #2646. https://github.com/bokeh/bokeh/pull/2646#issuecomment-133991602

My thought would be, add a new property on resources that returned the script so that, for example, you could do things like this:
````python
from bokeh.resources import CDN
from bokeh.embed import components
from jinja2 import Template  # This would also work for django template renderer
script, div = components(plot)
template = Template("""
<head>
    {{ CDN.css_resources.html }}
</head>
<body>
    {{ div }}
    {{ CDN.js_resources.html }}
    {{ script }}
</body>
""").render(script=script, resources=CDN, div=div)
````
or
````
from bokeh.resources import JSResources
from bokeh.embed import components
from jinja2 import Template 
script, div = components(plot)
template = Template("""
<head>
    <link href="my custom.css" type="text/css" rel="stylesheet" />
</head>
<body>
    {{ div }}
    {{ JSResources.html }}
    {{ script }}
</body>
""").render(script=script, resources=JSResources(mode='relative-dev'), div=div)
````

I think it'll also be interesting to visit this once the great split up is done as we think about how to comfortably provide different pieces to the user, so they can include them as needed.

**Comment by bryevdv** [comment]
> But if you put this in you would actually get a warning because resources argument isn't used and should be None.

The issue is older than the change that removed resources from components.

**Comment by birdsarah** [comment]
I'm guessing @chdoig would be okay with closing this. I don't think any of my `components` work has changed the issue.

**Comment by bryevdv** [comment]
I really like this idea:
```
 {{ CDN.css_resources.html }}
```
It could probably go in the other PR, with docs or examples?

**Comment by damianavila** [comment]
>I really like this idea:

ditto

**Comment by bbartoldson** [comment]
I liked @birdsarah's example above, but I couldn't get it working until I came across this demo: https://github.com/bokeh/bokeh/blob/master/examples/embed/simple/simple.py. Instead of using {{ CDN.css_resources.html }}, I used {{ INLINE.render_css() }} after importing INLINE from bokeh.resources.

**Comment by bryevdv** [comment]
Closing as very stale
